frame/staking-rewards/src/lib.rs

```diff
@@ -590,14 +601,21 @@ pub mod pallet {
 			// Transfer the reward locally.
 			let protocol_account = Self::account_id(asset);
 			T::Assets::transfer(*reward_asset, from, &protocol_account, amount, keep_alive)?;
-			EpochRewards::<T>::try_mutate(
-				(Self::current_epoch(), *asset),
-				reward_asset,
-				|collected_amount| -> DispatchResult {
-					*collected_amount = collected_amount.safe_add(&amount)?;
-					Ok(())
-				},
-			)?;
+			let f = |rewards: &mut RewardsOf<T>| -> DispatchResult {
+				let already_collected =
+					rewards.get(&reward_asset).copied().unwrap_or(BalanceOf::<T>::zero());
+				let new_collected = already_collected.safe_add(&amount)?;
+				rewards
+					.try_insert(*reward_asset, new_collected)
+					.map_err(|_| Error::<T>::TooManyRewardAssets)?;
+				Ok(())
+			};
+			match Self::current_state() {
+				State::WaitingForEpochEnd =>
+					EpochRewards::<T>::try_mutate(Self::current_epoch(), asset, f),
```

Comment:
This works for few things, I assume EpochRewards <-> PendingEpochRewards transfered back and forth, so if there will be 1000 assets it may take a while. So in general that not so good e.g. for positions.